src/types.jl

```diff
@@ -109,27 +128,50 @@ dimensions according to the operation.
 - `Quantity{T,D}(...)`: As above, but converting the value to type `T` and dimensions to `D`. You may also pass a
   `Quantity` as input.
 """
-struct Quantity{T,D<:AbstractDimensions} <: AbstractQuantity{T,D}
+struct Quantity{T<:Number,D<:AbstractDimensions} <: AbstractQuantity{T,D}
     value::T
     dimensions::D
 
     Quantity(x::_T, dimensions::_D) where {_T,_D<:AbstractDimensions} = new{_T,_D}(x, dimensions)
 end
-(::Type{Q})(x::T, ::Type{D}; kws...) where {D<:AbstractDimensions,T,T2,Q<:AbstractQuantity{T2}} = constructor_of(Q)(convert(T2, x), D(; kws...))
-(::Type{Q})(x, ::Type{D}; kws...) where {D<:AbstractDimensions,Q<:AbstractQuantity} = constructor_of(Q)(x, D(; kws...))
-(::Type{Q})(x::T; kws...) where {T,T2,Q<:AbstractQuantity{T2}} = constructor_of(Q)(convert(T2, x), dim_type(Q)(; kws...))
-(::Type{Q})(x; kws...) where {Q<:AbstractQuantity} = constructor_of(Q)(x, dim_type(Q)(; kws...))
 
-(::Type{Q})(q::AbstractQuantity) where {T,D<:AbstractDimensions,Q<:AbstractQuantity{T,D}} = constructor_of(Q)(convert(T, ustrip(q)), convert(D, dimension(q)))
-(::Type{Q})(q::AbstractQuantity) where {T,Q<:AbstractQuantity{T}} = constructor_of(Q)(convert(T, ustrip(q)), dimension(q))
+"""
+    GenericQuantity{T<:Any,D<:AbstractDimensions} <: AbstractGenericQuantity{T,D} <: Any
+
+This has the same behavior as `Quantity` but is subtyped to `AbstractGenericQuantity <: Any`
+rather than `AbstractQuantity <: Number`.
+"""
+struct GenericQuantity{T,D<:AbstractDimensions} <: AbstractGenericQuantity{T,D}
+    value::T
+    dimensions::D
+
+    GenericQuantity(x::_T, dimensions::_D) where {_T,_D<:AbstractDimensions} = new{_T,_D}(x, dimensions)
+end
+
+
+for (type, base_type) in ABSTRACT_QUANTITY_TYPES
+    @eval begin
+        (::Type{Q})(x::T, ::Type{D}; kws...) where {D<:AbstractDimensions,T<:$base_type,T2,Q<:$type{T2}} = constructor_of(Q)(convert(T2, x), D(; kws...))
+        (::Type{Q})(x::$base_type, ::Type{D}; kws...) where {D<:AbstractDimensions,Q<:$type} = constructor_of(Q)(x, D(; kws...))
+        (::Type{Q})(x::T; kws...) where {T<:$base_type,T2,Q<:$type{T2}} = constructor_of(Q)(convert(T2, x), dim_type(Q)(; kws...))
+        (::Type{Q})(x::$base_type; kws...) where {Q<:$type} = constructor_of(Q)(x, dim_type(Q)(; kws...))
+    end
+    for (type2, _) in ABSTRACT_QUANTITY_TYPES
```

Comment:
Can we skip this when `type2 = type1`?

Reply:
I don't think so, because these are abstract types. You would still want this to cover `MyQuantity1(::MyQuantity2)` even if those are **both** `AbstractQuantity`. Whether they are from different abstract types isn't important though